plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiMetadata.java

```diff
@@ -0,0 +1,336 @@
+import static org.apache.hudi.common.fs.FSUtils.getFs;
+import static org.apache.hudi.common.table.HoodieTableMetaClient.METAFOLDER_NAME;
+import static org.apache.hudi.exception.TableNotFoundException.checkTableValidity;
+
+public class HudiMetadata
+        implements ConnectorMetadata
+{
+    private static final Logger log = Logger.get(HudiMetadata.class);
+    private final HiveMetastore metastore;
+    private final HdfsEnvironment hdfsEnvironment;
+    private final TypeManager typeManager;
+
+    public HudiMetadata(HiveMetastore metastore, HdfsEnvironment hdfsEnvironment, TypeManager typeManager)
+    {
+        this.metastore = requireNonNull(metastore, "metastore is null");
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+    }
+
+    @Override
+    public List<String> listSchemaNames(ConnectorSession session)
+    {
+        return metastore.getAllDatabases();
+    }
+
+    @Override
+    public HudiTableHandle getTableHandle(ConnectorSession session, SchemaTableName tableName)
+    {
+        requireNonNull(tableName, "tableName is null");
+        if (isHiveSystemSchema(tableName.getSchemaName())) {
+            return null;
+        }
+        Optional<Table> table = metastore.getTable(tableName.getSchemaName(), tableName.getTableName());
+        if (table.isEmpty()) {
+            return null;
+        }
+        if (!isHudiTable(session, table.get())) {
+            throw new TrinoException(HUDI_UNKNOWN_TABLE_TYPE, format("Not a Hudi table: %s", tableName));
+        }
+        return new HudiTableHandle(
+                tableName.getSchemaName(),
+                tableName.getTableName(),
+                table.get().getStorage().getLocation(),
+                HoodieTableType.COPY_ON_WRITE,
+                TupleDomain.all(),
+                TupleDomain.all(),
+                Optional.of(getTableMetaClient(session, table.get())));
+    }
+
+    @Override
+    public ConnectorTableMetadata getTableMetadata(ConnectorSession session, ConnectorTableHandle table)
+    {
+        HudiTableHandle hudiTableHandle = (HudiTableHandle) table;
+        return getTableMetadata(hudiTableHandle.getSchemaTableName());
+    }
+
+    @Override
+    public Optional<ConstraintApplicationResult<ConnectorTableHandle>> applyFilter(ConnectorSession session, ConnectorTableHandle tableHandle, Constraint constraint)
+    {
+        HudiTableHandle handle = (HudiTableHandle) tableHandle;
+        HudiPredicates predicates = splitPredicate(constraint.getSummary());
+        HudiTableHandle newHudiTableHandle = handle.withPredicates(predicates);
+
+        if (handle.getPartitionPredicates().equals(newHudiTableHandle.getPartitionPredicates())
+                && handle.getRegularPredicates().equals(newHudiTableHandle.getRegularPredicates())) {
+            log.info("No new predicates to apply");
```

Comment:
this log is not actionable. Drop or move to `debug` adding extra information about the table.